src/events.c

```diff
@@ -0,0 +1,103 @@
+#include "events.h"
+
+/* this is only to queue GDK key events, in order to later send them if the map didn't match */
+static struct {
+    GdkEventKey **queue;                /* queue holding submitted events */
+    int            qlen;                /* pointer to last char in queue */
+    bool           processing;          /* whether or not events are processing */
+} events;
+
+/**
+ * Append an event into the queue.
+ */
+void queue_event(GdkEventKey* e)
+{
+    if (vb.mode->id != 'i') {
+        /* events are only needed for input mode */
+        return free_events();
+    }
+
+    GdkEventKey **newqueue = realloc(events.queue, (events.qlen + 1) * sizeof **newqueue);
+
+    if (newqueue == NULL) {
+        // error allocating memory
+        return;
+    }
+
+    events.queue = newqueue;
+
+    /* copy memory (otherwise event gets cleared by gdk) */
+    GdkEventKey* tmp = malloc(sizeof *tmp);
+    memcpy(tmp, e, sizeof *e);
+
+    if (tmp == NULL) {
+        // error allocating memory
+        return;
+    }
+
+    events.queue[events.qlen] = tmp;
+    events.qlen ++;
+}
+
+/**
+ * Free latest event & decrement qlen.
+ */
+void pop_event()
+{
+    if (events.qlen == 0) {
+        return;
+    }
+
+    free(events.queue[events.qlen - 1]);
+    events.qlen --;
+}
+
+/**
+ * Process events in the queue, sending the key events to GDK.
+ */
+void process_events(bool is_timeout)
```

Comment:
Added bool here since we need to case out the popping of events. Goal is to remove this in a future commit.